doc/distribution.md

```diff
@@ -0,0 +1,50 @@
+# Distribution
+
+This document outlines the expected way to distribute Ruby, with a specific focus on building Ruby packages for operating system distributions.
+
+## Building a Ruby Tarball
+
+The standard way to build a package for distribution is to build a tarball. This tarball includes all the related files need to build and install Ruby correctly. This includes the Ruby source code, the Ruby standard library, and the Ruby documentation.
+
+```bash
+$ ./autogen.sh
+$ ./configure -C
+$ make
```

Comment:
Building `ruby` from the source itself is not necessary, if `ruby` is available already.

Simply `ruby ./tool/make-snapshot tmp` can work.

Reply:
But most distributions build ruby in an environment where ruby is not yet available.

Reply:
Why needs to make tarballs in such environment, instead of installing?

Reply:
First of all, working `ruby` (currently >= 2.2) is required to build `ruby` from a git-cloned source tree, and `tool/make-snapshot` works only on a git-cloned source tree.
Although this sounds like chicken-and-egg, `make dist` is not intended for all Ruby developers, but for only the Ruby release managers.

Reply:
We use `https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.4.tar.xz` to build ruby from and it doesn't like like we need ruby for that.

Reply:
> But most distributions build ruby in an environment where ruby is not yet available.

In a downstream Fedora Ruby RPM, we are also building Ruby from the `ruby-*.tar.xz`  in an environment where the `ruby` is not available.

Reply:
Then I’m curious why this section is needed.

Reply:
I don't think we can assume that all distributions build it from the released tarballs, e.g. nix might conceivably build from source. We should outline the full process, from the source code through to making a package and installing it IMHO.

Reply:
Actually, this section is useful for building snapshot packages. And I am using tool/make-snapshot for that.

Reply:
@ioquatix I agree that it is a good idea, but it probably is not "for operating system distributions".

@voxik You're creating tarballs by yourself and then building packages?

Reply:
No, we don't. For official Fedora releases, we always use official tarballs. I don't remember any exception for past decade.

However, I am using self prepared tarballs for test builds, especially closer to the major upstream releases. I could use a nightly snapshots, but if there is some fast turnaround with fixes, the nightly snapshots are not flexible enough. And since I already have script which prepares the tarball as well as adjusts patches we are carrying around and updates part of the .spec file, I am using them always.

Reply:
This was very confusing, the first section of this document was basically something package managers should rarely if ever do (they should use release tarballs for packaging). I fixed it in https://github.com/ruby/ruby/commit/826067b0f970fe7d5514c908be5a30c308a836b5.

Reply:
Thanks!